src/basic_recipes/voronoiplot.jl

```diff
@@ -0,0 +1,166 @@
+                      smooth=false,
+
+                      # Point settings
+                      markersize=4,
+                      marker=sc.marker,
+                      point_color=sc.color,
+
+                      # Polygon settings
+                      strokecolor=theme(scene, :patchstrokecolor),
+                      strokewidth=1.0,
+                      polygon_color=automatic,
+                      unbounded_edge_extension_factor=0.1,
+                      bounding_box=automatic,
+
+                      # Colormap settings
+                      colormap=th.colormap,
+                      colorrange=th.colorrange,
+                      cycle=th.cycle)
+end
+
+function get_voronoi_tiles!(generators, polygons, vorn, bbox)
+    empty!(generators)
+    empty!(polygons)
+    sizehint!(generators, DelTri.num_generators(vorn))
+    sizehint!(polygons, DelTri.num_polygons(vorn))
+    if !isnothing(bbox)
+        a, b, c, d = bbox
+        @assert a < b && c < d "Bounding box must be of the form (xmin, xmax, ymin, ymax)."
+    end
+    for i in DelTri.each_generator(vorn)
+        g = DelTri.get_generator(vorn, i)
+        x, y = DelTri.getxy(g)
+        push!(generators, Point2f(x, y))
+        !isnothing(bbox) &&
+            @assert a ≤ x ≤ b && c ≤ y ≤ d "Generator $(i) with coordinates ($x, $y) is outside the bounding box."
+        polygon_coords = DelTri.get_polygon_coordinates(vorn, i, bbox)
+        polygon_coords_2f = map(polygon_coords) do coords
+            x, y = DelTri.getxy(coords)
+            !isnothing(bbox) &&
+                @assert a ≤ x ≤ b && c ≤ y ≤ d "Polygon vertex $(i) with coordinates ($x, $y) is outside the bounding box."
+            return Point2f(x, y)
+        end
+        push!(polygons, Polygon(polygon_coords_2f))
+    end
+    return generators, polygons
+end
+
+function get_voronoi_colors!(colors, vorn, cmap)
+    empty!(colors)
+    sizehint!(colors, DelTri.num_polygons(vorn))
+    F = DelTri.number_type(vorn)
+    gtr = [DelTri.get_generator(vorn, i) for i in DelTri.each_generator(vorn)]
+    reverse!(gtr) # For some reason this is needed to get distinct colors for the tiles
+    gtr_mat = reinterpret(reshape, F, gtr)
+    _colors = get(cgrad(cmap), gtr_mat, :extrema)
+    for c in eachcol(_colors)
+        a, b = c
+        push!(colors, (a + b) / 2)
+    end
+    return colors
+end
+
+
+Makie.conversion_trait(::Voronoiplot) = PointBased()
+Makie.convert_arguments(::Voronoiplot, x::DelTri.VoronoiTessellation) = (x,)
+
+function Makie.plot!(p::Voronoiplot{<: Tuple{<: Vector{<: Point2f}}})
+    attr = copy(p.attributes)
+    smooth = pop!(attr, :smooth)
+
+    # Handle transform_func early so tessellation is in cartesian space.
+    vorn = map(p.transformation.transform_func, p[1], smooth) do tf, ps, smooth
+        transformed = Makie.apply_transform(tf, ps)
+
+        # TODO: Make this work with Point2f directly
+        tri = DelTri.triangulate(Point2{Float64}.(transformed))
```

Comment:
I noticed that with Float32 this errors somewhere down the line. Would be good to fix this

Reply:
Can you share the stacktrace you saw? I thought I had fixed this previously. e.g.

```
julia> triangulate(rand(Float32, 2, 64))
Delaunay Triangulation.
    Constrained: false
    Has ghost triangles: false
    Number of points: 64
    Number of triangles: 116
    Number of edges: 189
```

```
julia> triangulate(Point2f[rand(2) for _ in 1:50])
Delaunay Triangulation.
    Constrained: false
    Has ghost triangles: false
    Number of points: 50
    Number of triangles: 83
    Number of edges: 147
```

Reply:
I thought I could get away without converting to a matrix but `centroidal_smooth` seems to error with points

Reply:
Ah. You are right:

```
julia> centroidal_smooth(voronoi(triangulate(pts)))
ERROR: MethodError: no method matching set_point!(::Vector{Point{2, Float32}}, ::Int64, ::Float32, ::Float32)

Closest candidates are:
  set_point!(::Any, ::Any, ::Any)
   @ DelaunayTriangulation C:\Users\User\.julia\packages\DelaunayTriangulation\Hlp1Y\src\interfaces\points.jl:290
  set_point!(::Triangulation, ::Any, ::Any, ::Any)
   @ DelaunayTriangulation C:\Users\User\.julia\packages\DelaunayTriangulation\Hlp1Y\src\data_structures\triangulation\points.jl:78
  set_point!(::AbstractMatrix{T}, ::Any, ::Any, ::Any) where T
   @ DelaunayTriangulation C:\Users\User\.julia\packages\DelaunayTriangulation\Hlp1Y\src\interfaces\points.jl:286
  ...

Stacktrace:
 [1] move_generator_to_centroid!(points::Vector{Point{…}}, vorn::VoronoiTessellation{Triangulation{…}, Tuple{…}, Int64, Tuple{…}, Set{…}, Tuple{…}}, generator::Int64)
   @ DelaunayTriangulation C:\Users\User\.julia\packages\DelaunayTriangulation\Hlp1Y\src\voronoi\lloyd.jl:7
 [2] _centroidal_smooth_itr(vorn::VoronoiTessellation{…}, set_boundary_nodes::Set{…}, points::Vector{…}, edges::Nothing, boundary_nodes::Vector{…}, I::Type, E::Type, V::Type, Es::Type, Ts::Type, F::Type, rng::Random.TaskLocalRNG; kwargs::@Kwargs{})
   @ DelaunayTriangulation C:\Users\User\.julia\packages\DelaunayTriangulation\Hlp1Y\src\voronoi\lloyd.jl:22
 [3] _centroidal_smooth_itr(vorn::VoronoiTessellation{…}, set_boundary_nodes::Set{…}, points::Vector{…}, edges::Nothing, boundary_nodes::Vector{…}, I::Type, E::Type, V::Type, Es::Type, Ts::Type, F::Type, rng::Random.TaskLocalRNG)
   @ DelaunayTriangulation C:\Users\User\.julia\packages\DelaunayTriangulation\Hlp1Y\src\voronoi\lloyd.jl:17
 [4] centroidal_smooth(vorn::VoronoiTessellation{Triangulation{…}, Tuple{…}, Int64, Tuple{…}, Set{…}, Tuple{…}}; maxiters::Int64, tol::Float64, rng::Random.TaskLocalRNG, kwargs::@Kwargs{})
   @ DelaunayTriangulation C:\Users\User\.julia\packages\DelaunayTriangulation\Hlp1Y\src\voronoi\main.jl:77
 [5] centroidal_smooth(vorn::VoronoiTessellation{Triangulation{…}, Tuple{…}, Int64, Tuple{…}, Set{…}, Tuple{…}})
   @ DelaunayTriangulation C:\Users\User\.julia\packages\DelaunayTriangulation\Hlp1Y\src\voronoi\main.jl:59
 [6] top-level scope
   @ REPL[15]:1
Some type information was truncated. Use `show(err)` to see complete types.
```

I've made the function `set_point!` too specific:

```julia
function set_point!(points::AbstractVector{T}, i, x, y) where {F,T<:NTuple{2,F}}
    points[i] = (F(x), F(y))
end
```

Probably needs to be something else like

```
function set_point!(points::AbstractVector{T}, i, x, y) where {T<:AbstractVector}
    points[i] = T(x, y) # or whatever the conversion needs to be ...
end
```

I have to come back and test this in DelaunayTriangulation.jl some more.

Reply:
I have fixed a bunch of DelaunayTriangulation.jl issues relating to Float32 (by just converting to Float64 internally - Float32 is really not sufficient for what is needed for computing triangulations/tessellations unfortunately, since ExactPredicates.jl is limited to Float64) (https://github.com/DanielVandH/DelaunayTriangulation.jl/pull/74 and https://github.com/DanielVandH/DelaunayTriangulation.jl/pull/75). Hopefully you don't run into any issues anymore.